lib/services.cf

```diff
@@ -125,6 +125,99 @@ body service_method standard_services
+      "non_disabling" or => { "start", "stop", "restart", "reload" };
+
+# A collection of classes to determine the capabilities of a given systemd
+# service, then start, stop, etc. the service. Also supports a custom action
+# for anything not supported
+
+      "service_enabled" expression => reglist(@(systemd_service_info), "UnitFileState=enabled");
+      "service_active"  expression => reglist(@(systemd_service_info), "ActiveState=active");
+      "service_loaded"  expression => reglist(@(systemd_service_info), "LoadState=loaded");
+      "service_notfound" expression => reglist(@(systemd_service_info), "LoadState=not-found");
+
+      "can_stop_service"   expression => reglist(@(systemd_service_info), "CanStop=yes");
+      "can_start_service"  expression => reglist(@(systemd_service_info), "CanStart=yes");
+      "can_reload_service" expression => reglist(@(systemd_service_info), "CanReload=yes");
+
+      "request_start"   expression => strcmp("start", "$(state)");
+      "request_stop"    expression => strcmp("stop", "$(state)");
+      "request_reload"  expression => strcmp("reload", "$(state)");
+      "request_restart" expression => strcmp("restart", "$(state)");
+
+      "action_custom"  expression => "!(request_start|request_stop|request_reload|request_restart)";
+      "action_start"   expression => "request_start.!service_active.can_start_service";
+      "action_stop"    expression => "request_stop.service_active.can_stop_service";
+      "action_reload"  expression => "request_reload.service_active.can_reload_service";
+      "action_restart"         or => {
+                                      "request_restart.service_active",
+
+                                      # Possibly undesirable... if a reload is
+                                      # requested, and the service "can't" be
+                                      # reloaded, then we restart it instead.
+                                      "request_reload.!can_reload_service.service_active",
+                                     };
+
+      # Starting a service implicitly enables it
+      "action_enable"  expression => "request_start.!service_enabled";
+
+      # Respectively, stopping it implicitly disables it
+      "action_disable" expression => "request_stop.service_enabled";
+
+  commands:
+    systemd.service_loaded:: # note this class is defined in `inventory/linux.cf`
+      # conveniently, systemd states map to `services` states, except
+      # for `enable`
+
+      "$(call_systemctl) -q start $(service)"   ifvarclass => "action_start";
+      "$(call_systemctl) -q stop $(service)"    ifvarclass => "action_stop";
+      "$(call_systemctl) -q reload $(service)"  ifvarclass => "action_reload";
+      "$(call_systemctl) -q restart $(service)" ifvarclass => "action_restart";
+      "$(call_systemctl) -q enable $(service)"  ifvarclass => "action_enable";
+      "$(call_systemctl) -q disable $(service)" ifvarclass => "action_disable";
```

Comment:
The promise will fail to evaluate if `$(state)` is not defined. The logic for class expressions is different but I'm pretty sure unexpanded variables are evaluation failures. As an edge case, I think this can be ignored, but we could enforce it with an explicit test like `if => some("^$(state)$", '["start", "stop", ...]')` (untested but you see what I mean).